src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalReadmeResource.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.Packaging.Core;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.Protocol
+{
+    /// <summary>
+    /// Finds the download url of a nupkg
+    /// </summary>
+    public class LocalReadmeResource : ReadmeResource
```

Comment:
I'm wondering if we need the lower level resource. 
Why is readme different from say icons, or licenses or other embedded content?
You'd normally have a response that would point to the location of the content, I don't see readme as special.

Now, I understand you probably have a reason that I'm missing, so I'd love to understand it better.

Reply:
I wrote a bit about it in my response to Andy here, https://github.com/NuGet/NuGet.Client/pull/5920#discussion_r1717654819. 

Looking at what we do the Icons and Licenses, I think that could work for the README, and it would go along with what Andy suggested, where add then new field "RawReadmeUrl" and for the local implementation it would be the location of the readme on disk. That said the implementation there doesn't do any caching for the HTTP results. 

My instinct is that the Icons and Licenses probably should move to a similar solution as the Readme, having Nuget.Protocol expose APIs for retrieving embedded content, instead of having clients implement solutions for themselves.

Reply:
An alternative, until downloading readmes from HTTP sources is implemented, is to avoid any changes to NuGet.Protocol, and just do the relevant work in `src\NuGet.Clients\*`. 

It looks to me like all the code in this class could be moved to `src\NuGet.Clients\*` without much modification. Then, once we also implement the APIs for downloading readmes from HTTP feeds, then we'll be able to make better public API decisions.

Reply:
From an architectural point of view, most resource classes tend to be associated with a completely new resource in the service index itself. 

Is there a specific resource for readme? 
To me it looks like readme is just metadata for an id, so it's not really a resource.

Reply:
That's true, I was wondering that. 

If we want to stick with that model, then we would either:
1. Modify the resources that map to the service index entries, add a new entry TryGetReadme
2. Follow what we do for the other embedded resources. Add RawReadmeUrl to the PackageMetadata, and let the embedded resource class figure out how to download stuff. 

My bias is still towards putting it in the resource so NuGet.Protocol is responsible for determining how get these resources, as opposed to the client. Though if we want to avoid committing to a public API right now, we could follow what's done for the other embedded resources and define the public API at a later date. It could even be defined after the remote feeds are implemented since the code for the other embedded resources has an HTTP client for fetching the data.

Reply:
Getting a readme isn't a resource in the same way getting an icon file isn't a resource. 
Adding it there pretty clearly violates that design pattern. 

NuGet.Protocol knows auth, so I think there's some need for these calls to be plumbed through NuGet.Protocol, but I just don't know if a resource is the right pattern.

Reply:
You mean that modifying the existing resources would violate that design pattern? I understand if we don't want a specific README one, but I think it could make sense to modify [PackageMetadataResource](https://github.com/NuGet/NuGet.Client/blob/71eb9731020bce05c3ef57f3a00a1dfcb504d6a1/src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResource.cs) to include a call to retrieve the README information. 

Could it make sense to create an "EmbeddedResource" which would handle all the embedded files and any new ones that we may want to selectively extract/download from the service?